### PR TITLE
Task/effective pipeline creation

### PR DESCRIPTION
**Pull request by esotrope**

### CODE REVIEW

File: script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java [hunk 1]
```diff
@@ -47,6 +50,10 @@ public PipelineDefinition getPipelineDefinition() {
         return pipelineDefinition;
     }
 
+    public List<ScriptDefinition> getScripts() {
+        return scriptDefinitions;
```
Comment (esotrope) [inline script/src/main/java/org/opentestsystem/rdw/script/Pipeline.java]: exposed this for testability.

@gregcharles one thing i noticed is we can simplify things to not use the "PipelineDefinition" at all i think. right now we map published pipeline to pipeline def and then never use what is on the pipeline def - only what is passed here as script defs. is that a correct observation?

File: script/src/main/java/org/opentestsystem/rdw/script/impl/DefaultPipelineFactory.java [hunk 2]
```diff
@@ -75,67 +75,56 @@ public DefaultPipelineFactory(final PublishedPipelineRepository repository,
...
-                                                "No base script found for the \"%s\" pipeline",
-                                                pipelineCode
-                                        )))
-                        ))
-                        .build()
+                createDevelopmentPipeline(pipelineCode, userScript)
         );
     }
 
+    private PublishedPipeline getEffectivePipeline(
```
Comment (esotrope) [inline script/src/main/java/org/opentestsystem/rdw/script/impl/DefaultPipelineFactory.java]: new logic block for getting the effective pipeline "def"

File: script/src/main/java/org/opentestsystem/rdw/script/publishing/ArchiveBackedPublishedPipelineRepository.java [hunk 3]
```diff
@@ -33,20 +30,18 @@
 @Repository
 public class ArchiveBackedPublishedPipelineRepository implements PublishedPipelineRepository {
 
+    private static final Pattern UserScriptPattern = Pattern.compile("^(.*?)user\\.(\\d+)\\.groovy$");
```
Comment (esotrope) [inline script/src/main/java/org/opentestsystem/rdw/script/publishing/ArchiveBackedPublishedPipelineRepository.java]: _more_ exact than `.groovy$` but is it exact enough @malaffoon ?

File: script/src/main/java/org/opentestsystem/rdw/script/publishing/PublishedPipelines.java [hunk 4]
```diff
@@ -8,6 +8,12 @@
  */
 public final class PublishedPipelines {
 
+    // Note the property keys had to be all lower case due to the way Amazon S3 coerces all property names to lower case
+    private static final String PipelineCodeProperty = "pipeline_code";
+    private static final String VersionProperty = "version";
+    private static final String PublishedOnProperty = "published_on";
+    private static final String PublishedByProperty = "published_by";
```
Comment (esotrope) [inline script/src/main/java/org/opentestsystem/rdw/script/publishing/PublishedPipelines.java]: lower case props

File: script/src/test/java/org/opentestsystem/rdw/script/PipelineFactoryTest.java [hunk 5]
```diff
@@ -71,54 +93,148 @@ public void before() throws Exception {
...
+        factory.getPipeline(publishedPipeline.getPipelineCode(), null);
+    }
+
+    @Test(expected = PipelineBuildException.class)
+    public void getPipelineShouldThrowExceptionWhenPostScriptDoesntCompile() {
+        when(scriptSource.loadScript(Post, publishedPipeline.getPipelineCode()))
+                .thenReturn(Optional.of(NonCompilingScript));
+
+        factory.getPipeline(publishedPipeline.getPipelineCode(), null);
+    }
```
Comment (esotrope) [inline script/src/test/java/org/opentestsystem/rdw/script/PipelineFactoryTest.java]: built out a bunch more tests for the pipeline factory - still need more

File: script/src/main/java/org/opentestsystem/rdw/script/publishing/ArchiveBackedPublishedPipelineRepository.java (hunk 3 above)
Comment (malaffoon) [reply script/src/main/java/org/opentestsystem/rdw/script/publishing/ArchiveBackedPublishedPipelineRepository.java]: Looks fine. Why the groupings, do we need them?

Comment (esotrope) [reply script/src/main/java/org/opentestsystem/rdw/script/publishing/ArchiveBackedPublishedPipelineRepository.java]: nope - will remove

**Pull request merged**